lib/services/ngsi/entities-NGSI-LD.js

```diff
@@ -383,6 +383,56 @@ function sendQueryValueNgsiLD(entityName, attributes, typeInformation, token, ca
     );
 }
 
+/**
+ * Adds any linked data items to the NGSI-LD payload for the context broker.
+ * For example a Building entity linked to a temperature gauge Device will
+ * also be updated as measures are received to update the Device Entity.
+ *
+ * @param {Object} typeInformation  Configuration information for the device.
+ * @param {Object} json             Payload to send to the context broker
+ */
+function addLinkedEntities(typeInformation, json) {
+    // use the supplied type if possible, but fallback to reading the
+    // type from the id of the entity.
+    function getEntityType(id, target) {
+        if (target.type) {
+            return target.type;
+        }
+        let calculatedType = id.replace(NGSI_LD_URN, '');
+        const n = calculatedType.indexOf(':');
+        calculatedType = calculatedType.substring(0, n !== -1 ? n : calculatedType.length);
+        return calculatedType;
+    }
+    const linkedTypeInfo = {};
+    if(typeInformation.staticAttributes){
+        typeInformation.staticAttributes.forEach((item) => {
+            if (item.link) {
+                linkedTypeInfo[item.name] = item.link;
+            }
+        });
+    }
+
+    Object.keys(linkedTypeInfo).forEach((sourceAttr) => {
+        const sourceAttribute = json[0][sourceAttr];
```

Comment:
are we guaranteed to find the Device payload in the 1st element of the array?

Reply:
For NGSI-LD, this is handled by `sendUpdateValueNgsiLD` - the existing call is an `upsert`:

```javascript
const url = '/ngsi-ld/v1/entityOperations/upsert/';

    const options = NGSIUtils.createRequestObject(url, typeInformation, token);
    options.method = 'POST';
```

The NGSI-LD Upsert is a batch operation and only accepts an array of elements, hence the 1st element must be a measure from a device.

Reply:
LGTM